### ISSUE TITLE
Support for Python 3.8 (broken?)

### ISSUE THREAD

**Issue by dataflake** [zopefoundation/ZEO#165]
Right now the ZEO package declares support for Python up to 3.7. Trying to run the tests on Python 3.8 produces two errors, see below. Basically, the `asyncio.Future` object `future` raises the exception because it is in status `cancelled`. The asynchronous operation "hides" the actual reason and I cannot figure it out.

Other than that there are some other failures on 3.8 that appear to be race and/or timing issues.

I cannot judge right now if running a ZEO server on Python 3.8 is "safe" in the context of e.g. a Zope 5 install under Python 3.8 or 3.9.

```
$ bin/test -tmultiple_addresses
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.


Error in test test_multiple_addresses (ZEO.asyncio.tests.ClientTests)
Traceback (most recent call last):
  File "/home/jens/src/Python-3.8.6/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/home/jens/src/Python-3.8.6/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/home/jens/src/Python-3.8.6/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/jens/src/ZEO/src/ZEO/asyncio/tests.py", line 430, in test_multiple_addresses
    self.respond(1, None)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/tests.py", line 119, in respond
    self.loop.protocol.data_received(
  File "/home/jens/src/ZEO/src/ZEO/asyncio/base.py", line 128, in data_received
    self.message_received(collected)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 232, in message_received
    future.set_result(args)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 904, in set_result
    self.cb(self)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 45, in _
    f = gen.send(result)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 202, in finish_connect
    self.client.registered(self, server_tid)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 447, in registered
    self._clear_protocols(protocol)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 404, in _clear_protocols
    p.close()
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 98, in close
    self._connecting.cancel()
  File "/home/jens/src/ZEO/src/ZEO/asyncio/testing.py", line 31, in call_soon
    func(*args)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 129, in done_connecting
    if future.exception() is not None:
asyncio.exceptions.CancelledError



Error in test test_multiple_addresses (ZEO.asyncio.tests.MsgpackClientTests)
Traceback (most recent call last):
  File "/home/jens/src/Python-3.8.6/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/home/jens/src/Python-3.8.6/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/home/jens/src/Python-3.8.6/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/jens/src/ZEO/src/ZEO/asyncio/tests.py", line 430, in test_multiple_addresses
    self.respond(1, None)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/tests.py", line 119, in respond
    self.loop.protocol.data_received(
  File "/home/jens/src/ZEO/src/ZEO/asyncio/base.py", line 128, in data_received
    self.message_received(collected)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 232, in message_received
    future.set_result(args)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 904, in set_result
    self.cb(self)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 45, in _
    f = gen.send(result)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 202, in finish_connect
    self.client.registered(self, server_tid)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 447, in registered
    self._clear_protocols(protocol)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 404, in _clear_protocols
    p.close()
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 98, in close
    self._connecting.cancel()
  File "/home/jens/src/ZEO/src/ZEO/asyncio/testing.py", line 31, in call_soon
    func(*args)
  File "/home/jens/src/ZEO/src/ZEO/asyncio/client.py", line 129, in done_connecting
    if future.exception() is not None:
asyncio.exceptions.CancelledError

  Ran 2 tests with 0 failures, 2 errors and 0 skipped in 0.038 seconds.
Tearing down left over layers:
  Tear down zope.testrunner.layer.UnitTests in 0.000 seconds.
```

**Comment by jugmac00** [comment]
I was not aware of this issue, and I upgraded the used Python version. I am running Python 3.8 with ZEO for a while and I did not notice any problems, but a broken test environment is concerning.

**Comment by vernans** [comment]
I tried to fix it the least invasive and correct way.
The problem is caused by this [python 3.8 change](https://bugs.python.org/issue32528).
Some context:
CancelledError inherits from "BaseException" now. This was somewhat of a reaction to people's habit to catch "Exception" in their Code and retry or abort quietly when working with  "Future" of concurrent.future. While this is possible for a broad selection of cases, CancelledError was often forgotten to handle seperately. It's an intended breaking change, while the way of breaking it seems odd to me. (choosing between "except BaseException:" or "except Exception:" doesn't seem very semantic to me, maybe someone knows more.)
To make things more weird, we also have a seperate "class Fut(object):" which is a modified implementation of Future. It handles exceptions inheriting "Exception" differently (than concurrent.future.Future). Fut throws the Exception, concurrent.future.Future logs and ignores them (while it throws BaseException).
In "test_multiple_addresses" there actually occured two bugs. One: it is possible to have a CancelledError while closing a protocol (when trying to cancel the pending Future). Two: In earlier versions (incorrect behavior because catching CancelledError) the second protocol wasn't even tried to be closed. If it was closed (by properly catching this Error), it would have crashed in a ClientDisconnected error in this test. 
This is what I believe to know, feel free to adapt the code.
By "least invasive" i mean, it would have also been possible to catch this error at different locations. (some of the except Exception: clauses would have been possible), but I think maybe we shouldn't restore the old behavior if it is explicitly discouraged.

**Comment by dataflake** [comment]
I agree that we shouldn't have code that is explicitly discouraged. I also like how your changes are small.

**Comment by ctheune** [comment]
@vernans as I just stumbled over this in a different piece of software and you asked about the semantics: regular application code would use the idiom of 'except Exception' to handle all application-level exceptions. However, a long time ago, this used to be 'except:' which has been strongly discouraged due to it interfering with system-level exceptions that indicate stuff like the user pressing Ctrl-C and where you want a developer to make a very conscious decision whether to suppress it or not. Having the CancelledError inherit from the regular Exception will cause problems with exactly this kind of signalling on a framework level and cause asyncio programs to not shut down properly in a lot of situations. 

I think the CancelledError handling needs to be taught more as it's quite relevant to writing proper asyncio software on a basic level. I just discovered it last week myself, using 3.7 I think, and immediately wanted the CancelledError to be a BaseException, so I was happy to see that this was improved already.
